### ISSUE TITLE
Seed data observations outside the simulation spatial domain cause extract operator error

### ISSUE THREAD

**Issue by bryce-carson** [ashokkrish/spatialEpisim#5]
While attempting to replicate the issue in #4 as described by Ashok in a personal communication, I encountered the following error (which causes the usual busy client UX).

```
 grc_ppp_2020_1km_Aggregated_UNadj
 Min.   :    0.0                  
 1st Qu.:    2.5                  
 Median :    5.7                  
 Mean   :   49.9                  
 3rd Qu.:   15.7                  
 Max.   :18873.2                  
 NA's   :820316                   
[1] 820316
Warning: Error in h: error in evaluating the argument 'i' in selecting a method for function '[': NA/NaN argument
  1: runApp
```

**Comment by bryce-carson** [comment]
The issue is not present if the province of _Crete_ is not cropped from the data when running a default SVEIRD simulation.

**Comment by bryce-carson** [comment]
```
 grc_ppp_2020_1km_Aggregated_UNadj
 Min.   :    0.0                  
 1st Qu.:    2.5                  
 Median :    5.7                  
 Mean   :   49.9                  
 3rd Qu.:   15.7                  
 Max.   :18873.2                  
 NA's   :820316                   
[1] 820316
[1] "Susceptible Count before removing initial seed values:  9033978.84338977"
[1] "Susceptible Count after removing initial seed values:  9033978.84338977"
[1] "time =  1"
[1] "time =  2"
[1] "time =  3"
[1] "time =  4"
[1] "time =  5"
[1] "time =  6"
[1] "time =  7"
[1] "time =  8"
[1] "time =  9"
[1] "time =  10"
Output #0, mp4, to 'Infected_MP4.mp4':
  Metadata:
    encoder         : Lavf60.16.100
  Stream #0:0: Video: mpeg4 (mp4v / 0x7634706D), yuv420p, 768x768, q=2-31, 200 kb/s, 90k tbn
    Side data:
      cpb: bitrate max/min/avg: 0/0/200000 buffer size: 0 vbv_delay: N/A
Adding frame 11 at timestamp 15.00sec (90%) - video stream completed!
[1] "isCropped" "TRUE"     
Warning: Error in h: error in evaluating the argument 'i' in selecting a method for function '[': NA/NaN argument
  1: runApp
```

**Comment by ashokkrish** [comment]
For all of Greece could set the rasterAgg to a small number (say 5, 10) and re-run?

**Comment by bryce-carson** [comment]
The same issue occurs with the simulation options in #8, but cropping the _default_ Congolese provinces.

**Comment by bryce-carson** [comment]
> For all of Greece could set the rasterAgg to a small number (say 5, 10) and re-run?

I will try that; determining if the aggregation factor is related may be important.

**Comment by bryce-carson** [comment]
> For all of Greece could set the rasterAgg to a small number (say 5, 10) and re-run?

The issue occurs with that aggregation factor as well.

**Comment by bryce-carson** [comment]
# Diagnosis
**_Cropping the province containing the only seed data_** is the likely source of the issue.

All the following locations are on the island of Crete:

| Location       | lat    | lon    | InitialVaccinated | InitialExposed | InitialInfections | InitialRecovered | InitialDead |
|----------------|--------|--------|-------------------|----------------|-------------------|------------------|-------------|
| Chania         | 35.517 | 24.017 | 1000              | 200            | 100               | 10               | 10          |
| Heraklion      | 35.34  | 24.134 | 1000              | 200            | 100               | 10               | 10          |
| Rethymno       | 35.369 | 24.474 | 1000              | 200            | 100               | 10               | 10          |
| Ierapetra      | 35.012 | 25.741 | 1000              | 200            | 100               | 10               | 10          |
| Agios Nikolaos | 35.183 | 25.712 | 1000              | 200            | 100               | 10               | 10          |
| Sitia          | 35.2   | 26.1   | 1000              | 200            | 100               | 10               | 10          |

![](https://upload.wikimedia.org/wikipedia/commons/1/18/Crete_integrated_map-en.svg)

# Discussion
@ashokkrish, in your email asking for me to prioritize this issue this week, you said

> I ran an SVEIRD model simulation for 50 days **_in the island of Crete_** in Greece using the above seed data.

You wrote as if you intend for the simulation to be _limited to the island of Crete_, when what is happening when we crop this island is the simulation is running _everywhere in Greece but Crete_. The seed data is limited to Crete, so there is nothing for the modelling process to work with when Crete is cropped from the data.

# Related issues
Why the issue #4 occurs when the simulation is set to Greece will be determined in #4.

# Solution
The simulation can reject the seed data if there is any data point in it which occurs in a cropped province or state. An alternative is to crop such observations from the seed data, and also ensuring that the final seed data has a final length of at least one.

```R
## Ensure there is at least one observation in the seed data.
stopifnot(length(filteredSeedData))
```

**Comment by bryce-carson** [comment]
I don't know why Ashok was able to run the simulation, cropping the island of Crete, in #4. Perhaps he used different seed data?

**Comment by bryce-carson** [comment]
I can replicate this issue when using [this seed data *COD_InitialSeedData_Beni.csv*](https://github.com/ashokkrish/spatialEpisim/blob/9cf5016b3d1c1509488102e6865c1db78e6d7566/seeddata/COD_InitialSeedData_Beni.csv).

Other options:
- SVEIRD
- Crop map to / limit simulation to _Nord Kivu_ (North Kivu)

**Comment by bryce-carson** [comment]
**_Without cropping the map / limiting the simulation to_** Nord Kivu—i.e. `isCropped = FALSE` is logged—the simulation will run with the same initial seed data.

**Comment by bryce-carson** [comment]
**Without cropping,** the leaflet plot is automatically reducing the zoomed portion to Ituri and Nord Kivu. What are the defaults in the application? Is it automatically cropping the simulation to Ituri and Nord Kivu?

![Screen Shot 2024-06-11 at 21 43 27](https://github.com/ashokkrish/spatialEpisim/assets/51974245/dcdaa5b5-c798-459a-b039-2c1a802cf5e1)
![Screen Shot 2024-06-11 at 21 43 17](https://github.com/ashokkrish/spatialEpisim/assets/51974245/50ead26a-fb37-44bf-a021-5a055e9c2787)

**Comment by ashokkrish** [comment]
> **_Without cropping the map / limiting the simulation to_** Nord Kivu—i.e. `isCropped = FALSE` is logged—the simulation will run with the same initial seed data.

Yes it would run as long as a seed data is uploaded. The seed data may be for the whole country or selected set of cities.

What's problematic is if the data is seeded outside the spatial domain. I assume the simulation runs even if you select only one province however the seed data is for two provinces. Correct?

**Comment by ashokkrish** [comment]
> **Without cropping,** the leaflet plot is automatically reducing the zoomed portion to Ituri and Nord Kivu. What are the defaults in the application? Is it automatically cropping the simulation to Ituri and Nord Kivu?
> 
> ![Screen Shot 2024-06-11 at 21 43 27](https://private-user-images.githubusercontent.com/51974245/338815629-dcdaa5b5-c798-459a-b039-2c1a802cf5e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxNjQzMTgsIm5iZiI6MTcxODE2NDAxOCwicGF0aCI6Ii81MTk3NDI0NS8zMzg4MTU2MjktZGNkYWE1YjUtYzc5OC00NTlhLWIwMzktMmMxYTgwMmNmNWUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEyVDAzNDY1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzZDVjNmYxNzEwZWEwODAwNzdmOGQ1ZTIxMzU1NWE4MmIzNzA4OGVmYTE0ZjExMzE4NTk3NmE2NTg1YmEyM2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.EDQxod0L5GpB3G3LsHJ-WLY4O8YbyF0eIEluhC4EUdI) ![Screen Shot 2024-06-11 at 21 43 17](https://private-user-images.githubusercontent.com/51974245/338815634-50ead26a-fb37-44bf-a021-5a055e9c2787.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxNjQzMTgsIm5iZiI6MTcxODE2NDAxOCwicGF0aCI6Ii81MTk3NDI0NS8zMzg4MTU2MzQtNTBlYWQyNmEtZmIzNy00NGJmLWEwMjEtNWEwNTVlOWMyNzg3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEyVDAzNDY1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEzMzM4YWQzNmQ3MzFlYThmN2E2MDZmNjM1Y2ExNTVhMmI4Y2NjNjdiODI3ZTJhNDU4YmU4ZDRhYTJkODkzNDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PorR_I2KU4m30R2I4z1GRwpM9Uo6HrwV8BXX6VFg2cc)

Without cropping it runs for the whole country for me.

**Comment by ashokkrish** [comment]
> I don't know why @ashokkrish was able to run the simulation, cropping the island of Crete, in #4. Perhaps he used different seed data?

I used GRC_Crete_InitialSeedData.csv and I made sure I cropped Crete on the left panel.

**Comment by bryce-carson** [comment]
> > **_Without cropping the map / limiting the simulation to_** Nord Kivu—i.e. `isCropped = FALSE` is logged—the simulation will run with the same initial seed data.
> 
> Yes it would run as long as a seed data is uploaded. The seed data may be for the whole country or selected set of cities.
> 
> What's problematic is if the data is seeded outside the spatial domain.

I encounter _this issue's error_ when I run a simulation of Greece limited to Athos with the Crete data. Perhaps the issue is caused by the presence of data outside of the domain. More testing will be needed, because this is a difficult to locate error; maybe I'll have better luck running the application in showcase mode to see exactly what is running before the error is encountered; it happens very quickly, however. I'll check in on Shiny debugging methodology first.

**Comment by bryce-carson** [comment]
I said,
> I can replicate this issue when using [this seed data _COD_InitialSeedData_Beni.csv_](https://github.com/ashokkrish/spatialEpisim/blob/9cf5016b3d1c1509488102e6865c1db78e6d7566/seeddata/COD_InitialSeedData_Beni.csv).
> 
> Other options:
> 
>     * SVEIRD
> 
>     * Crop map to / limit simulation to _Nord Kivu_ (North Kivu)

You replied to another thing I said,

> > **_Without cropping the map / limiting the simulation to_** Nord Kivu—i.e. `isCropped = FALSE` is logged—the simulation will run with the same initial seed data.
> 
> Yes it would run as long as a seed data is uploaded. The seed data may be for the whole country or selected set of cities.
> 
> What's problematic is if the data is seeded outside the spatial domain. I assume the simulation runs even if you select only one province however the seed data is for two provinces. Correct?

It took me about fifteen minutes to definitively determine where the Manima housing development is, but I saw on the map that it's next to Mambasa and that's definitely within the Ituri province, so the seed data does cover two provinces.

Running a simulation with the Beni seed data and limiting the simulation to _either_ Ituri or Nord Kivu will cause this error, but if I include both provinces the simulation will run.

# Conclusion
The cause of this issue, determined through scientific deduction 🧑‍🔬 🧪, is the presence of seed data outside of the simulation domain. _Any_ seed data outside of the simulation domain seems to be an issue. The solution is, then, as I said earlier but with the assumption that cropping worked in the opposite direction to how it does work in the application, to ensure that no point of seed data exists outside of the polygons.

That won't be too difficult because the vector data of a coordinate and a polygon can definitely be quickly checked. We can ensure that all points of seed data exist within _some_ polygon the simulation is including, and if not prevent the simulation from running.

**Comment by bryce-carson** [comment]
> > The solution is, then, as I said earlier but with the assumption that cropping worked in the opposite direction to how it does work in the application, to ensure that no point of seed data exists outside of the polygons.
> > That won't be too difficult because the vector data of a coordinate and a polygon can definitely be quickly checked. We can ensure that all points of seed data exist within _some_ polygon the simulation is including, and if not prevent the simulation from running.
> 
> I have wanted this kind of validation for a long time now (almost two years) but told to myself hopefully the user uploads the propre seed data. Yes, we will address in the coming updates. Just wanted to caution you about not getting too deep into this as I would like a stable version by this Friday to get my slides going.

Indeed. I'll work on the feature _in the unstable branch_ that I'll create. The stable branch (main) has now received the fixes for the other identified issues we discussed yesterday.

These are the items on my checklist I wrote yesterday during our meeting which I have not yet worked on:
- [ ] Run a Campania, Italy simulation
- [ ] Validate uploaded seed data is non-negative
- [ ] Edit seed data templates so all compartments are positive

It's possible to refactor the UI code and make some edits to the server without upsetting the current logic (which would also resolve #18), but I'll leave that to the unstable branch just so that not too much changes in the code and your familiarity with it isn't undermined.

**Comment by bryce-carson** [comment]
Here are resources I'll be using to resolve this issue, which is in the _point in polygon_ domain of spatial data analysis.

- https://stackoverflow.com/a/66382301/14211497
- https://observablehq.com/@tmcw/understanding-point-in-polygon
- https://en.wikipedia.org/wiki/Point_in_polygon

StackOverflow user Robert Hijmans provides four examples using the _terra_ library.

**Comment by ashokkrish** [comment]
This is excellent. Please go ahead.

**Comment by Toby-exe** [comment]
@ashokkrish please see the latest commit (61564e2ad12f56b29bffd3fd48e957f66c2276d5) for progress on this issue

**Comment by Toby-exe** [comment]
I believe this issue can be closed as of 321ef2b0b465a81f3d265d1cb5f29e1cad400939
